scripts/xe-restore-metadata

```diff
@@ -110,16 +122,75 @@ sr_name=$(${XE} sr-param-get uuid="${sr_uuid}" param-name=name-label)
 # probe first for a VDI with known UUID derived from the SR to avoid
 # scanning for a VDI
 backup_vdi=$(uuid5 "${NS}" "${sr_uuid}")
-if [ -z "${vdis}" ]; then
-  vdis=$(${XE} vdi-list uuid="${backup_vdi}" sr-uuid="${sr_uuid}" read-only=false --minimal)
+
+# Only allow a specified VDI that does not match the known UUID if operating in
+# legacy mode
+if [ -n "${specified_vdi}" ]; then
+  if [ "${specified_vdi}" != "${backup_vdi}" ] && [ "$legacy" -eq 0 ]; then
+    echo "The specified VDI UUID is not permitted, if attempting to use a legacy backup VDI please use the -o flag" >&2
+    exit 1
+  fi
+  vdis=${specified_vdi}
 fi
 
-# get a list of all VDIs if an override has not been provided on the cmd line
 if [ -z "${vdis}" ]; then
-  vdis=$(${XE} vdi-list params=uuid sr-uuid="${sr_uuid}" read-only=false --minimal)
+  if [ "$legacy" -eq 0 ]; then
```

Comment:
Do we need legacy mode support on master? Would the code be simpler if we removed it?
We didn't have this at all at XS8 release time I think, so you couldn't have created a legacy mode backup on XS8.
(well you could have created one on an earlier release but that means rolling back a VM to a previous release after potentially some changes were already done to it on XS8)

Reply:
Unfortunately we do need it, as you can pull a backup from 8.2 (indeed for some users it is how they want to upgrade - unplug SR, do new installs on all the hosts, then use this feature to recreate their VMs)

Reply:
You're right.
The legacy form is also the only one that works on SMAPIv3 (there is no API to introduce deterministic UUIDs there, it gets silently dropped by `xapi-storage-script` and a random UUID used instead).
So for GFS2 and 4KSR support we probably need to retain the legacy mode for a while (unless we make the necessary changes to xapi-storage-script and the storage plugins to support it).